src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodImportCompletionHelper.ExtensionMethodSymbolComputer.cs

```diff
@@ -0,0 +1,539 @@
+            private ImmutableArray<string> ReceiverTypeNames { get; }
+            private ISet<string> NamespaceInScope { get; }
+            private ConcurrentDictionary<ITypeSymbol, bool> CheckedReceiverTypes { get; }
+            private IImportCompletionCacheService<CacheEntry, object> CacheService { get; }
+
+            public ExtensionMethodSymbolComputer(
+                Document document,
+                SemanticModel semanticModel,
+                ITypeSymbol receiverTypeSymbol,
+                ImmutableArray<string> receiverTypeNames,
+                int position,
+                ISet<string> namespaceInScope,
+                ConcurrentDictionary<ITypeSymbol, bool> checkedReceiverTypes,
+                IImportCompletionCacheService<CacheEntry, object> cacheService)
+            {
+                CurrentDocument = document;
+                SemanticModel = semanticModel;
+                ReceiverTypeSymbol = receiverTypeSymbol;
+                ReceiverTypeNames = receiverTypeNames;
+                Position = position;
+                NamespaceInScope = namespaceInScope;
+                CheckedReceiverTypes = checkedReceiverTypes;
+                CacheService = cacheService;
+            }
+
+            public static async Task<ExtensionMethodSymbolComputer> CreateAsync(
+                Document document,
+                int position,
+                ITypeSymbol receiverTypeSymbol,
+                ISet<string> namespaceInScope,
+                CancellationToken cancellationToken)
+            {
+                var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+                // This dictionary is used as cache among all projects and PE references. 
+                // The key is the receiver type as in the extension method declaration (symbol retrived from current compilation).
+                // The value indicates if we can reduce an extension method with this receiver type given receiver type.
+                var checkedReceiverTypes = new ConcurrentDictionary<ITypeSymbol, bool>();
+                var receiverTypeNames = GetReceiverTypeNames(receiverTypeSymbol);
+                receiverTypeNames = AddComplexTypes(receiverTypeNames);
+                var cacheService = GetCacheService(document.Project.Solution.Workspace);
+
+                return new ExtensionMethodSymbolComputer(
+                    document, semanticModel, receiverTypeSymbol, receiverTypeNames, position, namespaceInScope, checkedReceiverTypes, cacheService);
+            }
+
+            /// <summary>
+            /// Force create all relevant indices
+            /// </summary>
+            public Task PopulateIndicesAsync(CancellationToken cancellationToken)
+            {
+                using var _ = ArrayBuilder<Task>.GetInstance(out var tasks);
+
+                foreach (var project in GetAllRelevantProjects(CurrentProject))
```

Comment:
is GetAllRelevantProjects always passed CurrentProject?  if so it doesn't need to be passed.   if not, disregard this question.